### ISSUE TITLE
TypeError: Cannot read property 'forEach' of undefined

### ISSUE THREAD

**Issue by sahariko** [webpack-contrib/webpack-bundle-analyzer#490]
#### Issue description

When running the CLI on a stats file (not all stats files, just some), I'm getting the following error, and the following error only:

```
Could't analyze webpack bundle:
TypeError: Cannot read property 'forEach' of undefined
```

It's not a lot to go on, there aren't any helpful hints or file names in which it crashes.

#### Technical info

<pre>
System:
    OS: macOS 11.6
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 3.08 GB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 12.22.0 - ~/.nvm/versions/node/v12.22.0/bin/node
    Yarn: 1.22.17 - ~/.nvm/versions/node/v12.22.0/bin/yarn
    npm: 6.14.15 - ~/.nvm/versions/node/v12.22.0/bin/npm
  npmPackages:
    terser-webpack-plugin: 5.1.3 => 5.1.3 
    webpack: 5.64.1 => 5.64.1 
    webpack-cli: 4.7.2 => 4.7.2 
    webpack-dev-server: 4.7.3 => 4.7.3 
    webpack-manifest-plugin: 3.1.1 => 3.1.1 
</pre>

#### Debug info

Using the CLI - `npx webpack-bundle-analyzer@latest stats.json`, with the following plugins:
- [`@sentry/webpack-plugin`](https://github.com/getsentry/sentry-webpack-plugin)
- [`terser-webpack-plugin`](https://webpack.js.org/plugins/terser-webpack-plugin/)
- [`webpack-manifest-plugin`](https://github.com/shellscape/webpack-manifest-plugin)
- [`mini-css-extract-plugin`](https://github.com/webpack-contrib/mini-css-extract-plugin)
- [`terser-webpack-plugin`](https://webpack.js.org/plugins/terser-webpack-plugin/)
- [`webpack.ContextReplacementPlugin`](https://webpack.js.org/plugins/context-replacement-plugin/)
- [`webpack.DefinePlugin`](https://webpack.js.org/plugins/define-plugin/)

**Comment by valscion** [comment]
> on a stats file (not all stats files, just some),

Are you able to post a stats file that causes this issue?

It's possible that some presence checks are needed where we use `.forEach`. It's possible that when usage of lodash was made smaller, there now is less protection against some data missing and `.forEach` still being called on them:

* https://github.com/webpack-contrib/webpack-bundle-analyzer/pull/395

**Comment by sahariko** [comment]
I don't think it would be possible, since it's a proprietary codebase 😕

**Comment by sahariko** [comment]
But if it helps in any way we're exporting a function that returns an array.

When returning an array we're getting `.filter` errors, when returning an object (Webpack config), we're getting the `.forEach` error.

If I had to take a shot I would assume these lines are problematic:

https://github.com/webpack-contrib/webpack-bundle-analyzer/pull/395/files#diff-cc250993338f2c875d952782d3e6e5465c510ee355a68adb9f8f1ccd5d446823L51

https://github.com/webpack-contrib/webpack-bundle-analyzer/pull/395/files#diff-cc250993338f2c875d952782d3e6e5465c510ee355a68adb9f8f1ccd5d446823L210

**Comment by valscion** [comment]
We can't fix it unless we can make a test case. Otherwise we're likely to break the fix in a future release if we don't have test coverage.

**Comment by sahariko** [comment]
I completely understand, but I think that more explicit errors can be very helpful, for example check if `bundleStats.assets` is defined, and if it isn't print a more helpful error message, maybe with a possible reason as to why it's undefined.

```
TypeError: Cannot read property 'forEach' of undefined
```

Is very cryptic, and doesn't even help me understand if there's an issue with this package, my config, or something else.

**Comment by valscion** [comment]
That is an error that merely comes because something is `undefined` and then `.forEach` is getting called on it. We need to know what is undefined to fix it, or to even add a better error message if we can't fix it.

Without a stats JSON file, we don't know what is undefined and we can't write tests that would ensure we don't regress any error message feature or a test fix in the future.

**Comment by ilias-t** [comment]
Here's a `compilation-stats.json` file that repros the issue (apologies for it being rather large)
[compilation-stats.json.zip](https://github.com/webpack-contrib/webpack-bundle-analyzer/files/7973733/compilation-stats.json.zip)

**Comment by thecristen** [comment]
hi. here's a stats.json file that produces this error, using latest webpack, webpack-cli, webpack-bundle-analyzer:

```
Could't analyze webpack bundle:
TypeError: Cannot read property 'filter' of undefined
```

[stats.json.zip](https://github.com/webpack-contrib/webpack-bundle-analyzer/files/8486062/stats.json.zip)

**Comment by sebastian-nowak** [comment]
Any updates on this? This is still an issue in the latest lib version, and it makes it unusable.

**Comment by sahariko** [comment]
@valscion Don't know about a fix, but both @ilias-t and @thecristen provided Zip files with a reproducible error.

**Comment by valscion** [comment]
Yes, I askes about a reproduction as a test case and creating a pull request that aims to fix this bug.

I don't have time right now to tackle this and it seems that this isn't affecting too many people right now. The quickest way to get this fixed is to help in fixing it as part of open source ☺️

**Comment by Ismaestro** [comment]
Ok, so as an alternative, use source-map-explorer...

```
# package.json scripts section

"bundle-report": "ng build --configuration=production-sourcemaps && source-map-explorer dist/angularexampleapp/browser/**/*.js"
```

```
# angular.json

"configurations": {
   ...
   "production-sourcemaps": {
      "sourceMap": true,
      "outputHashing": "all"
   },
   ...
}
```

**Comment by hecerinc** [comment]
FYI, I dug into the code a little bit, and it was analyzer.js:getViewerData() function returning the undefined in this chunk:

```js
			children[i].assets.forEach((asset) => {
				asset.isChild = true;
				bundleStats.assets.push(asset);
			});

```
I added the null coalescing operator and then it failed again. Upon inspection it looks like `bundleStats.assets` was empty, so I figured it was missing a property in the stats when running `webpack --profile --json`, so I changed the profiling command to `webpack --profile --json --stats verbose`, and it started working 🚀

**Comment by valscion** [comment]
Thanks for the insight @hecerinc! I wonder if there's some place in the README.md here where this `--stats verbose` would be useful to highlight? If so, feel free to send a PR for said update ☺️
